src/cli/cli_mesh_diag.cpp

```diff
@@ -0,0 +1,543 @@
+ */
+
+#include "cli_mesh_diag.hpp"
+
+#include <openthread/mesh_diag.h>
+
+#include "cli/cli.hpp"
+#include "cli/cli_utils.hpp"
+#include "common/code_utils.hpp"
+
+#if OPENTHREAD_CONFIG_MESH_DIAG_ENABLE
+
+namespace ot {
+namespace Cli {
+
+MeshDiag::MeshDiag(otInstance *aInstance, OutputImplementer &aOutputImplementer)
+    : Utils(aInstance, aOutputImplementer)
+{
+}
+
+template <> otError MeshDiag::Process<Cmd("topology")>(Arg aArgs[])
+{
+    /**
+     * @cli meshdiag topology
+     * @code
+     * meshdiag topology
+     * id:02 rloc16:0x0800 ext-addr:8aa57d2c603fe16c ver:4 - me - leader
+     *    3-links:{ 46 }
+     * id:46 rloc16:0xb800 ext-addr:fe109d277e0175cc ver:4
+     *    3-links:{ 02 51 57 }
+     * id:33 rloc16:0x8400 ext-addr:d2e511a146b9e54d ver:4
+     *    3-links:{ 51 57 }
+     * id:51 rloc16:0xcc00 ext-addr:9aab43ababf05352 ver:4
+     *    3-links:{ 33 57 }
+     *    2-links:{ 46 }
+     * id:57 rloc16:0xe400 ext-addr:dae9c4c0e9da55ff ver:4
+     *    3-links:{ 46 51 }
+     *    1-links:{ 33 }
+     * Done
+     * @endcode
+     * @par
+     * Discover network topology (list of routers and their connections).
+     * Parameters are optional and indicate additional items to discover. Can be added in any order.
+     * * `ip6-addrs` to discover the list of IPv6 addresses of every router.
+     * * `children` to discover the child table of every router.
+     * @par
+     * Information per router:
+     * * Router ID
+     * * RLOC16
+     * * Extended MAC address
+     * * Thread Version (if known)
+     * * Whether the router is this device is itself (`me`)
+     * * Whether the router is the parent of this device when device is a child (`parent`)
+     * * Whether the router is `leader`
+     * * Whether the router acts as a border router providing external connectivity (`br`)
+     * * List of routers to which this router has a link:
+     *   * `3-links`: Router IDs to which this router has a incoming link with link quality 3
+     *   * `2-links`: Router IDs to which this router has a incoming link with link quality 2
+     *   * `1-links`: Router IDs to which this router has a incoming link with link quality 1
+     *   * If a list if empty, it is omitted in the out.
+     * * If `ip6-addrs`, list of IPv6 addresses of the router
+     * * If `children`, list of all children of the router. Information per child:
+     *   * RLOC16
+     *   * Incoming Link Quality from perspective of parent to child (zero indicates unknown)
+     *   * Child Device mode (`r` rx-on-when-idle, `d` Full Thread Device, `n` Full Network Data, `-` no flags set)
+     *   * Whether the child is this device itself (`me`)
+     *   * Whether the child acts as a border router providing external connectivity (`br`)
+     * @cparam meshdiag topology [@ca{ip6-addrs}] [@ca{children}]
+     * @sa otMeshDiagDiscoverTopology
+     */
+    otError error = OT_ERROR_NONE;
+
+    otMeshDiagDiscoverConfig config;
```

Comment:
Remove newline between local var declarations?
```suggestion
    otError                  error = OT_ERROR_NONE;
    otMeshDiagDiscoverConfig config;
```